clang/lib/CIR/CodeGen/CIRGenExprConst.cpp

```diff
@@ -377,15 +377,40 @@ mlir::Attribute ConstantAggregateBuilder::buildFrom(
   CharUnits AlignedSize = Size.alignTo(Align);
 
   bool Packed = false;
-  ArrayRef<mlir::Attribute> UnpackedElems = Elems;
+  ArrayRef<mlir::Attribute> UnpackedElems;
+
+  // Fullill the init elements for union. This comes from a fundamental
+  // difference between CIR and LLVM IR. In LLVM IR, the union is simply a
+  // struct with the largest member. So it is fine to have only one init
+  // element. But in CIR, the union has the information for all members. So if
+  // we only pass a single init element, we may be in trouble. We solve the
+  // problem by appending placeholder attribute for the uninitialized fields.
```

Comment:
This solution works fine, but I wonder if you have explored something more terse (e.g. adding a `union_index` attribute to `cir.const_struct` that makes the verifier assert for only one element and a valid index - this is kinda of a half baked idea to illustrate the point).

Reply:
I didn't try other solutions. The idea to add  `union_index` attribute to `cir.const_struct` sounds good. But I feel it is more or less a grammar sugar. I feel it didn't add new semantics than the current one. And given the consumer of the union can assert the type of the union arbitrarily, I am not sure what can we verify.

Reply:
Yea, definitely grammar sugar to make things look nicer, but we don't need to block progress with this right now, we can revisit if we find this to be polluting the CIR output in the future.